File: python/media_storage.py
"""Download media files from Telegram to local storage at archive time."""
import os
import logging
from config import MEDIA_STORAGE_DIR, MEDIA_MAX_DOWNLOAD_SIZE

logger = logging.getLogger(__name__)


def _find_cached(file_unique_id: str) -> str | None:
    """Return path if a file with this unique_id is already stored."""
    if not file_unique_id:
        return None
    try:
        for name in os.listdir(MEDIA_STORAGE_DIR):
            if name.startswith(file_unique_id):
                return os.path.abspath(os.path.join(MEDIA_STORAGE_DIR, name))
    except OSError:
        pass
    return None


def download_and_save(bot, file_id: str, file_unique_id: str,
                      file_size: int | None = None,
                      suggested_ext: str = "") -> str | None:
    """
    Download a Telegram file via Bot API and save to local storage.
    Returns absolute path on success, None on failure or skip.
    """
    if not file_id or not file_unique_id:
        return None

    # Already cached?
    existing = _find_cached(file_unique_id)
    if existing:
        return existing

    # Telegram Bot API limit is ~20 MB. Skip larger files.
    if file_size and file_size > MEDIA_MAX_DOWNLOAD_SIZE:
        logger.info(f'Skipping download of {file_unique_id}: size {file_size} > limit {MEDIA_MAX_DOWNLOAD_SIZE}')
        return None

    try:
        file_info = bot.get_file(file_id)
    except Exception as e:
        logger.warning(f'get_file failed for {file_id}: {e}')
        return None

    # Determine extension
    _, ext = os.path.splitext(file_info.file_path or "")
    if not ext and suggested_ext:
        ext = suggested_ext if suggested_ext.startswith('.') else '.' + suggested_ext

    out_path = os.path.abspath(os.path.join(MEDIA_STORAGE_DIR, file_unique_id + ext))

    try:
        data = bot.download_file(file_info.file_path)
        with open(out_path, 'wb') as f:
            f.write(data)
        logger.debug(f'Saved media to {out_path} ({len(data)} bytes)')
        return out_path
    except Exception as e:
        logger.warning(f'download failed for {file_id}: {e}')
        return None

File: rust/src/media_storage.rs
use std::path::{Path, PathBuf};
use log::{debug, warn};
use teloxide::prelude::*;
use teloxide::net::Download;
use tokio::fs;
use tokio::io::AsyncWriteExt;

/// Returns existing cached path if a file with this unique_id already exists.
async fn find_cached(storage_dir: &str, file_unique_id: &str) -> Option<PathBuf> {
    let mut entries = match fs::read_dir(storage_dir).await {
        Ok(e) => e,
        Err(_) => return None,
    };
    while let Ok(Some(entry)) = entries.next_entry().await {
        if let Some(name) = entry.file_name().to_str() {
            if name.starts_with(file_unique_id) {
                let path = entry.path();
                return std::fs::canonicalize(&path).ok().or(Some(path));
            }
        }
    }
    None
}

fn to_absolute(p: &Path) -> PathBuf {
    if p.is_absolute() {
        p.to_path_buf()
    } else {
        std::env::current_dir()
            .map(|cwd| cwd.join(p))
            .unwrap_or_else(|_| p.to_path_buf())
    }
}

/// Download a Telegram file via Bot API and save to storage_dir.
/// Returns absolute path on success, None on failure or skip.
pub async fn download_and_save(
    bot: &Bot,
    storage_dir: &str,
    file_id: &str,
    file_unique_id: &str,
    file_size: Option<i64>,
    max_size: u64,
) -> Option<String> {
    if file_id.is_empty() || file_unique_id.is_empty() {
        return None;
    }

    // Already cached?
    if let Some(p) = find_cached(storage_dir, file_unique_id).await {
        return p.to_str().map(|s| s.to_string());
    }

    // Skip files exceeding Bot API limit (~20 MB).
    if let Some(sz) = file_size {
        if sz as u64 > max_size {
            debug!("Skipping download of {}: size {} > limit {}", file_unique_id, sz, max_size);
            return None;
        }
    }

    // Resolve file path via getFile
    let tg_file = match bot.get_file(file_id).await {
        Ok(f) => f,
        Err(e) => {
            warn!("get_file failed for {}: {}", file_id, e);
            return None;
        }
    };

    // Pick extension from telegram-provided path
    let ext = Path::new(&tg_file.path)
        .extension()
        .and_then(|e| e.to_str())
        .map(|e| format!(".{}", e))
        .unwrap_or_default();

    let out_path = to_absolute(&PathBuf::from(storage_dir).join(format!("{}{}", file_unique_id, ext)));

    let mut file = match fs::File::create(&out_path).await {
        Ok(f) => f,
        Err(e) => {
            warn!("failed to create {}: {}", out_path.display(), e);
            return None;
        }
    };

    if let Err(e) = bot.download_file(&tg_file.path, &mut file).await {
        warn!("download failed for {}: {}", file_id, e);
        let _ = fs::remove_file(&out_path).await;
        return None;
    }
    let _ = file.flush().await;

    debug!("Saved media to {}", out_path.display());
    out_path.to_str().map(|s| s.to_string())
}
